### ISSUE TITLE
TerraClient prematurely raises TerraTimeoutException

### ISSUE THREAD

**Issue by hannes-ucsc** [DataBiosphere/azul#5682]
https://ucsc-gi.slack.com/archives/C03TPJS54DC/p1699400036142029

> 
> 
> Hi folks I am seeing this error frequently , immediately on entering the site any ideas?
> Error Message
> Request failed with status code 503 - No response from https://data.terra.bio/api/repository/v1/snapshots/roleMap within 5.0 seconds
> But it’s immediate. No 5 seconds required.
> Seems to get better if I just refresh but may come back if I’m idle for a while. (I’m not 100% sure on the idle for a while part)
> We can retry these, but I wanted to raise this to see if something else should be done . (edited) 

The key observation is that Azul raises this 503 very quickly, way before the 5s timeout is over. Note the latency column in the screenshots below. 

This starts at different times in anvilprod and prod, suggesting that it is not caused by a change to Terra since both deployments hit TDR prod.

anvilprod:

<img width="1292" alt="Screen Shot 2023-11-07 at 5 48 11 PM" src="https://github.com/DataBiosphere/azul/assets/5143256/a28e4453-57e3-47a4-ab35-afeea235b782">

prod:

<img width="1279" alt="Screen Shot 2023-11-07 at 5 48 53 PM" src="https://github.com/DataBiosphere/azul/assets/5143256/a4b6f4c5-7460-46e2-8e13-e3e812427748">

```
fields @timestamp, integrationLatency
| filter integrationLatency < 100 and status = 503
| sort @timestamp asc
| limit 10
```

The first occurrence in `anvilprod` correlates with deploy job [14533](https://gitlab.prod.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/14533) for commit 533fc45d5812c5e6a1ba6552056a0858c131eff8, specifically [this change](https://github.com/DataBiosphere/azul/commit/533fc45d5812c5e6a1ba6552056a0858c131eff8#diff-cbdc3142f0e57896cee357536c684df369dbbb302c775465a0e51627e8063086) to the retry policy.

The first occurrence in `prod`correlates with the promotion. 

- [x] Security design review completed; the Resolution of this issue does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

**Comment by hannes-ucsc** [comment]
```diff
Index: src/azul/terra.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/terra.py b/src/azul/terra.py
--- a/src/azul/terra.py	(revision 7e5b5470b150c19a6920314903233c920feb9454)
+++ b/src/azul/terra.py	(date 1699409090886)
@@ -344,18 +344,17 @@
         try:
             # Limited retries on I/O errors such as refused or dropped
             # connections. The latter are actually very likely if connections
-            # from the pool are reused after a long period of idleness.
+            # from the pool are reused after a long period of idleness. That's
+            # why we need at least one retry on read.
             retry = urllib3.Retry(total=None,
                                   connect=retries,
-                                  read=0,
+                                  read=retries + 1,
                                   redirect=0,
                                   status=0,
                                   other=retries)
             response = self._http_client.request(method,
                                                  str(url),
                                                  headers=headers,
-                                                 # FIXME: Service should return 503 response when Terra client times out
-                                                 #        https://github.com/DataBiosphere/azul/issues/3968
                                                  timeout=timeout,
                                                  retries=retry,
                                                  body=body)
```

The removal of the stale FIXME should happen in a separate commit whose title should reference the now closed #3968.

**Comment by hannes-ucsc** [comment]
For demo, show absence of matching log entries using CW Insights query from description.
